File: go2shop-cart-service/src/main/java/com/go2shop/cart/service/ShoppingCartService.java
package com.go2shop.cart.service;

import java.util.List;
import java.util.Optional;

import com.go2shop.common.exception.BusinessException;
import com.go2shop.model.cart.ShoppingCartDTO;
import com.go2shop.model.cart.ShoppingCartProductDTO;

public interface ShoppingCartService {
	
	ShoppingCartDTO createShoppingCart(ShoppingCartDTO shoppingCartDTO);
	
	Optional<ShoppingCartProductDTO> getShoppingCartProduct(Long shoppingCartProductID) throws BusinessException;
	
	List<ShoppingCartProductDTO> getAllShoppingCartProduct(Long shoppingCartID) throws BusinessException;
	
	void deleteAllProduct(Long shoppingCartID);
	
	void deleteShoppingCartProductById(Long shoppingCartProductId);
	
	void deleteShoppingCartProductByProductId(Long productId);
	
	void deleteShoppingCartProductByCartIdAndProductId(Long cartId, Long productId);
	
	ShoppingCartDTO getShoppingCartByUserId(Long userID);
	
	ShoppingCartProductDTO createShoppingCartProduct(ShoppingCartProductDTO shoppingCartProductDTO);
	
	ShoppingCartProductDTO updateQuantity(Long cartId, Long productID, int productQuantity);
	
	Long countCartSizeByDistinctProductIds(Long userId);
}

File: go2shop-app/src/app/pages/cart/shopping-cart.service.ts
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable, Subject } from 'rxjs';
import { IShoppingCartProduct, ShoppingCartProduct } from './shopping-cart-product.model';
import { IShoppingCart } from './shopping-cart.model';

@Injectable({
  providedIn: 'root'
})
export class ShoppingCartService {
    private updateCartSizeSubj = new Subject();
    updateCartSizeObs = this.updateCartSizeSubj.asObservable();

    private buyNowSubj = new BehaviorSubject<boolean>(false);
    buyNowObs = this.buyNowSubj.asObservable();
    buyNowProduct: ShoppingCartProduct;
    
    private api = 'api/cartService/shoppingCart';

    constructor(private http: HttpClient) {}

    updateShoppingCartState(isBuyNow: boolean, product?: ShoppingCartProduct): void {
      this.buyNowSubj.next(isBuyNow);
      this.buyNowProduct = product;
    }

    updateCartSize(): void {
      this.updateCartSizeSubj.next();
    }

    getCartSize(userId: number): Observable<HttpResponse<number>> {
      return this.http.get<number>(`${this.api}/cart/size/${userId}`, { observe: 'response' });
    }

    createShoppingCart(shoppingCartDTO: IShoppingCart): Observable<HttpResponse<IShoppingCart>> {
      return this.http.post<IShoppingCart>(`${this.api}/shoppingCart/create`, shoppingCartDTO,
        { observe: 'response' });
    }

    createShoppingCartProduct(shoppingCartProductDTO: IShoppingCartProduct): Observable<HttpResponse<IShoppingCartProduct>> {
      return this.http.post<IShoppingCartProduct>(`${this.api}/shoppingCartProduct/create`, shoppingCartProductDTO, 
        { observe: 'response' });
    }

    getShoppingCart(userID: number): Observable<HttpResponse<IShoppingCart>> {
      return this.http.get<IShoppingCart>(`${this.api}/shoppingCart/${userID}`, { observe: 'response' });
    }

    getShoppingCartProduct(shoppingCartProductID: number): Observable<HttpResponse<IShoppingCartProduct>> {
      return this.http.get<IShoppingCartProduct>(`${this.api}/shoppingCartProduct/product/${shoppingCartProductID}`, 
        { observe: 'response' });
    }

    getAllShoppingCartProduct(shoppingCartID: number): Observable<HttpResponse<IShoppingCartProduct[]>> {
      return this.http.get<IShoppingCartProduct[]>(`${this.api}/shoppingCartProduct/shoppingCart/${shoppingCartID}`, 
        { observe: 'response' });
    }

    deleteAllProduct(shoppingCartID: number): Observable<any> {
      return this.http.post<void>(`${this.api}/shoppingCartProduct/delete`, shoppingCartID,
        { observe: 'response' });
    }

    deleteShoppingCartProduct(shoppingCartProductID: number): Observable<any> {
      return this.http.post<void>(`${this.api}/shoppingCartProduct/delete/${shoppingCartProductID}`,
        { observe: 'response' });
    }

    updateQuantity(productID: number, productQuantity: number, shoppingCartID: number): Observable<HttpResponse<IShoppingCartProduct>> {
      return this.http.put<IShoppingCartProduct>
      (`${this.api}/shoppingCartProduct/update/${productID}/${shoppingCartID}/${productQuantity}`, 
        {productID, productQuantity, shoppingCartID}, 
        {observe: 'response'});
    }
}